src/synth/fluid_chan.c

```diff
@@ -294,3 +314,360 @@ fluid_channel_get_sfont_bank_prog(fluid_channel_t* chan, int *sfont,
+ *
+ *                The monophonic list
+ *  +------------------------------------------------+
+ *  | +--------------------------------------------+ |
+ *  | |  +----+   +----+          +----+   +----+  | |
+ *  | +--|note|<--|note|<--....<--|note|<--|note|<-+ |
+ *  +--->|vel |-->|vel |-->....-->|vel |-->|vel |----+
+ *       +----+   +----+          +----+   +----+
+ *        /|\                      /|\
+ *         |                        |
+ *      i_first                   i_last
+ * 
+ * @param chan  fluid_channel_t.
+ * @param key MIDI note number (0-127) to search.
+ * @return index of the note if find, INVALID_NOTE otherwise.
+ * 
+ */
+unsigned short
+fluid_channel_search_monolist(fluid_channel_t* chan, unsigned char key)
+{
+	short n = chan->n_notes; /* number of notes in monophonic list */
+	short i= chan->i_first; /* searching starts from i_first included */
+	while(n) 
+	{
+		if(chan->monolist[i].note == key) break; /* found */
+		i = chan->monolist[i].next; /* next element */
+		n--;
+	}
+	if (n)
+	{
+		return i;/* found i */
+	}
+	else
+	{
+		return INVALID_NOTE; /* not found */
+	}
+}
+
+/**
+ * removes a note out of the monophonic list.The function is part of 
+ * the legato detector. 
+ * fluid_channel_remove_monolist() is intended to be called by 
+ * fluid_synth_noteoff_mono_LOCAL().
+ *
+ *            The monophonic list
+ *  +------------------------------------------------+
+ *  | +--------------------------------------------+ |
+ *  | |  +----+   +----+          +----+   +----+  | |
+ *  | +--|note|<--|note|<--....<--|note|<--|note|<-+ |
+ *  +--->|vel |-->|vel |-->....-->|vel |-->|vel |----+
+ *       +----+   +----+          +----+   +----+
+ *        /|\                      /|\
+ *         |                        |
+ *      i_first                   i_last
+ *
+ * The monophonic list is a circular buffer of  SIZE_MONOLIST elements
+ * Each element is linked forward and backward at initialisation time.
+ * when a note is removed at noteOff the element concerned is fast unlinked
+ * and relinked after the i_last element.
+ *
+ * @param chan  fluid_channel_t.
+ * @param 
+ *   i, index of the note to remove. If i is invalid or the list is 
+ *      empty, the function do nothing and returns INVALID_NOTE.
+ * @return prev index prior the last note if i is the last note in the list,
+ *        INVALID_NOTE otherwise. When the returned index is valid it means
+ *        a legato dectection.
+ *
+ * Note: the following variables in Channel keeps trace of the situation.
+ *       - i_last index keeps a trace of the most recent note played even if
+ *       the list is empty.
+ *       - prev_note keeps a trace of the note removed if it is i_last.
+ *       - FLUID_CHANNEL_LEGATO_PLAYING bit keeps a trace of legato/staccato playing state.
+ * 
+ * More informations in FluidPolyMono-0003.pdf chapter 4 (Appendices).
+ */
+unsigned char
+fluid_channel_remove_monolist(fluid_channel_t* chan, short i)
```

Comment:
Same here. If this function returns an index I would prefer returning an `int` or `FLUID_FAILED`.

Reply:
Ok.